File: src/handlers/get_photos.py
"""Lambda handler for retrieving photos.

Supports listing all photos or filtering by a single tag. Returns metadata
together with time-limited presigned download URLs.
"""

import json
import os

import boto3
from boto3.dynamodb.conditions import Key

PHOTOS_TABLE = os.environ["PHOTOS_TABLE"]
PHOTOS_BUCKET = os.environ["PHOTOS_BUCKET"]

dynamodb = boto3.resource("dynamodb")
s3 = boto3.client("s3")
table = dynamodb.Table(PHOTOS_TABLE)

DEFAULT_LIMIT = 20
MAX_LIMIT = 100
PRESIGNED_URL_EXPIRY = 3600  # 1 hour


def handler(event, context):
    """Handle GET /photos requests."""
    params = event.get("queryStringParameters") or {}

    tag = params.get("tag")
    limit = _parse_limit(params.get("limit"))
    next_token = params.get("nextToken")

    if tag:
        photos, new_next_token = _query_by_tag(tag, limit, next_token)
    else:
        photos, new_next_token = _scan_all_photos(limit, next_token)

    # Enrich each photo with a presigned download URL
    for photo in photos:
        photo["downloadUrl"] = s3.generate_presigned_url(
            "get_object",
            Params={"Bucket": PHOTOS_BUCKET, "Key": photo["s3Key"]},
            ExpiresIn=PRESIGNED_URL_EXPIRY,
        )
        del photo["s3Key"]  # no need to expose internal key
        # Remove DynamoDB key attributes from the response
        photo.pop("PK", None)
        photo.pop("SK", None)

    response_body = {"photos": photos}
    if new_next_token:
        response_body["nextToken"] = new_next_token

    return _response(200, response_body)


def _query_by_tag(tag, limit, next_token):
    """Query the tag index to find photos matching a specific tag."""
    # Step 1: get photoIds from the TAG# partition
    query_args = {
        "KeyConditionExpression": Key("PK").eq(f"TAG#{tag}"),
        "Limit": limit,
    }
    if next_token:
        query_args["ExclusiveStartKey"] = json.loads(
            _b64_decode(next_token),
        )

    result = table.query(**query_args)
    tag_items = result.get("Items", [])

    # Step 2: batch-get the full metadata for each photo
    photos = []
    for item in tag_items:
        photo_id = item["photoId"]
        meta = table.get_item(
            Key={"PK": f"PHOTO#{photo_id}", "SK": "METADATA"},
        ).get("Item")
        if meta:
            photos.append(meta)

    new_next_token = None
    if "LastEvaluatedKey" in result:
        new_next_token = _b64_encode(json.dumps(result["LastEvaluatedKey"]))

    return photos, new_next_token


def _scan_all_photos(limit, next_token):
    """Scan for all photo metadata items."""
    scan_args = {
        "FilterExpression": Key("SK").eq("METADATA"),
        "Limit": limit,
    }
    if next_token:
        scan_args["ExclusiveStartKey"] = json.loads(
            _b64_decode(next_token),
        )

    result = table.scan(**scan_args)
    photos = result.get("Items", [])

    new_next_token = None
    if "LastEvaluatedKey" in result:
        new_next_token = _b64_encode(json.dumps(result["LastEvaluatedKey"]))

    return photos, new_next_token


# -- helpers ----------------------------------------------------------------

def _parse_limit(raw):
    try:
        value = int(raw)
        return max(1, min(value, MAX_LIMIT))
    except (TypeError, ValueError):
        return DEFAULT_LIMIT


def _b64_encode(text):
    import base64
    return base64.urlsafe_b64encode(text.encode()).decode()


def _b64_decode(token):
    import base64
    return base64.urlsafe_b64decode(token.encode()).decode()


def _response(status_code, body):
    return {
        "statusCode": status_code,
        "headers": {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
        },
        "body": json.dumps(body),
    }

File: lambda/get-photos.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  QueryCommand,
  QueryCommandInput,
  ScanCommand,
  ScanCommandInput,
  BatchGetCommand,
} from "@aws-sdk/lib-dynamodb";
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

const DEFAULT_LIMIT = 20;
const MAX_LIMIT = 100;
const PRESIGNED_URL_EXPIRY = 3600; // 1 hour

const dynamodb = DynamoDBDocumentClient.from(new DynamoDBClient({}));
const s3 = new S3Client({});

export async function handler(
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> {
  const PHOTOS_TABLE = process.env.PHOTOS_TABLE!;
  const PHOTOS_BUCKET = process.env.PHOTOS_BUCKET!;

  const params = event.queryStringParameters || {};

  const tag = params.tag;
  const limit = parseLimit(params.limit);
  const nextToken = params.nextToken;

  let photos: Record<string, unknown>[];
  let newNextToken: string | undefined;

  if (tag) {
    ({ photos, nextToken: newNextToken } = await queryByTag(
      PHOTOS_TABLE,
      tag,
      limit,
      nextToken,
    ));
  } else {
    ({ photos, nextToken: newNextToken } = await scanAllPhotos(
      PHOTOS_TABLE,
      limit,
      nextToken,
    ));
  }

  // Enrich each photo with a presigned download URL
  for (const photo of photos) {
    photo.downloadUrl = await getSignedUrl(
      s3,
      new GetObjectCommand({
        Bucket: PHOTOS_BUCKET,
        Key: photo.s3Key as string,
      }),
      { expiresIn: PRESIGNED_URL_EXPIRY },
    );
    delete photo.s3Key;
    delete photo.PK;
    delete photo.SK;
  }

  const responseBody: Record<string, unknown> = { photos };
  if (newNextToken) {
    responseBody.nextToken = newNextToken;
  }

  return response(200, responseBody);
}

async function queryByTag(
  tableName: string,
  tag: string,
  limit: number,
  nextToken?: string,
): Promise<{ photos: Record<string, unknown>[]; nextToken?: string }> {
  // Step 1: get photoIds from the TAG# partition
  const queryInput: QueryCommandInput = {
    TableName: tableName,
    KeyConditionExpression: "PK = :pk",
    ExpressionAttributeValues: { ":pk": `TAG#${tag}` },
    Limit: limit,
  };

  if (nextToken) {
    queryInput.ExclusiveStartKey = JSON.parse(b64Decode(nextToken));
  }

  const result = await dynamodb.send(new QueryCommand(queryInput));
  const tagItems = result.Items || [];

  // Step 2: batch-get the full metadata for each photo
  let photos: Record<string, unknown>[] = [];
  if (tagItems.length > 0) {
    const keys = tagItems.map((item) => ({
      PK: `PHOTO#${item.photoId}`,
      SK: "METADATA",
    }));

    const batchResult = await dynamodb.send(
      new BatchGetCommand({
        RequestItems: {
          [tableName]: { Keys: keys },
        },
      }),
    );
    photos = (batchResult.Responses?.[tableName] || []) as Record<
      string,
      unknown
    >[];
  }

  let newNextToken: string | undefined;
  if (result.LastEvaluatedKey) {
    newNextToken = b64Encode(JSON.stringify(result.LastEvaluatedKey));
  }

  return { photos, nextToken: newNextToken };
}

async function scanAllPhotos(
  tableName: string,
  limit: number,
  nextToken?: string,
): Promise<{ photos: Record<string, unknown>[]; nextToken?: string }> {
  const scanInput: ScanCommandInput = {
    TableName: tableName,
    FilterExpression: "SK = :sk",
    ExpressionAttributeValues: { ":sk": "METADATA" },
    Limit: limit,
  };

  if (nextToken) {
    scanInput.ExclusiveStartKey = JSON.parse(b64Decode(nextToken));
  }

  const result = await dynamodb.send(new ScanCommand(scanInput));
  const photos = (result.Items || []) as Record<string, unknown>[];

  let newNextToken: string | undefined;
  if (result.LastEvaluatedKey) {
    newNextToken = b64Encode(JSON.stringify(result.LastEvaluatedKey));
  }

  return { photos, nextToken: newNextToken };
}

// -- helpers ----------------------------------------------------------------

function parseLimit(raw?: string): number {
  const value = parseInt(raw || "", 10);
  if (isNaN(value)) return DEFAULT_LIMIT;
  return Math.max(1, Math.min(value, MAX_LIMIT));
}

function b64Encode(text: string): string {
  return Buffer.from(text).toString("base64url");
}

function b64Decode(token: string): string {
  return Buffer.from(token, "base64url").toString();
}

function response(
  statusCode: number,
  body: Record<string, unknown>,
): APIGatewayProxyResult {
  return {
    statusCode,
    headers: {
      "Content-Type": "application/json",
      "Access-Control-Allow-Origin": "*",
    },
    body: JSON.stringify(body),
  };
}
